### ISSUE TITLE
Thoughts on a ClusterAlertManagerConfig?

### ISSUE THREAD

**Issue by endzyme** [prometheus-operator/prometheus-operator#6919]
### Component(s)

AlertManager, AlertManagerConfig

### What is missing? Please describe.

I'd love to have a chat about supporting a `ClusterAlertManagerConfig` resource. While having an `AlertManagerConfig` has been amazing, tying common alerting rules to a group of namespaces makes for a lot of duplicate resources in many namespaces. 

Here's an amorphous list of thoughts of things this would probably impact:

- How would route ordering work when you have many ClusterAlertManagerConfigs and AlertManagerConfigs?
    - Might impact (#5781)
- Could you use receivers from ClusterAlertManagerConfigs in AlertManagerConfigs?
    - Would it be worth splitting out _yet another_ resource like `ClusterAlertManagerReceivers` to address the needs outlined in (#5653)
- Would ClusterAlertManagerConfigs just generate AlertManagerConfigs under the hood, using namespace selectors; or would it be better to have a separate implementation?
- What are the guardrails you want in place? It feels like there is one right now that `AlertManagerConfigs` will never collide because they're scoped to a namespace and _must_ have receivers defined in the same resource. How might these design decisions change with a cluster wide resource? Should it be a concern of this operator to "fix" some of the difficulties of working with the Alertmanager route tree? 

### Describe alternatives you've considered.

We thought about replicating or wrapping with a helm chart for some of our practices, but it felt like something the upstream should have and everything else would be a workaround. 

## Environment
Kubernetes Version:
Prometheus-Operator Version:

**Comment by simonpasquier** [comment]
Thanks for raising the issue! I'm not sure to understand exactly what you're asking for. Are you looking for splitting your Alertmanager configuration into multiple resources that the operator would collate? It would be good if you could share more details about your use case/environment.

**Comment by endzyme** [comment]
> Thanks for raising the issue! I'm not sure to understand exactly what you're asking for. Are you looking for splitting your Alertmanager configuration into multiple resources that the operator would collate? It would be good if you could share more details about your use case/environment.

Sure no problem!

Let me try to explain my use case a little more. 

I want to be able to create routes without being limited to a single namespace. My use case is that we have several apps that span many namespaces and it would be nice to organize their routes based on more than just namespace. 

For example, currently if I define the following resource:

```yaml
apiVersion: monitoring.coreos.com/v1alpha1
kind: AlertmanagerConfig
metadata:
  name: test-remove-me
spec:
  receivers:
    - name: 'null'
    - name: test-1
  route:
    continue: false
    receiver: test-1
```

I will see the following in the Alertmanager configuration:

```text
route:
  receiver: "null"
  continue: false
  routes:
  - receiver: default/test-remove-me/test-1
    matchers:
    - namespace="default"
    continue: true
receivers:
- name: "null"
- name: default/test-remove-me/null
- name: default/test-remove-me/test-1
```

It would be really nice if we were not constrained by the namespace here. It I wanted my matcher to be `namespace=~"default|other-namespace|third-ns"` then that would be really helpful. The only other alternative is creating a new AlertmanagerConfig in each of the namespaces I want to have routes for, which duplicate the receivers and all the routes. 

It would be really nice to be able to use either a namespace selector or just apply my own matchers _without_ the forced namespace part.

**Comment by simonpasquier** [comment]
Thanks for the details, it helped! Adding another CRD to the operator is a big endeavor for the team and something that we can't commit on lightly.

As you've described initially, the tricky part when collating different AlertmanagerConfig objects is making that they don't collide and that alerts are processed by all route entries.

On the Prometheus side, we already have the `.spec.excludedFromEnforcement` field which works alongside  `.spec.enforcedNamespaceLabel`: any monitor/rule object which is listed here (selected either by exact name or namespace) doesn't have their `namespace` label value enforced. It's an option we could explore for AlertmanagerConfig too.

**Comment by nicolastakashi** [comment]
I tend to agree with you @simonpasquier seems to me the limitation is only the default enforcement of the namespace on the alert routing matcher.

@endzyme do you think if we keep the default behaviour and add the ability to excludedFromEnforcement could work on your use case?

**Comment by endzyme** [comment]
Thank you @simonpasquier and @nicolastakashi for taking the time to answer.

I think that having an `enforcedNamespaceLabel` field would work for me and this use case. If you did use the approach of having `excludedFromEnforcement` on the `AlertmanagerConfig` then that could also affect the forced `continue: true` at the root level of the routes. 

In any case, I would love to have a field of `enforcedNamespaceLabel: false`. Is this something you'd want in a PR or does the project take on its own tasks like this one?

**Comment by simonpasquier** [comment]
You can already use https://prometheus-operator.dev/docs/api-reference/api/#monitoring.coreos.com/v1.AlertmanagerConfigMatcherStrategy

````
spec:
  alertmanagerConfigMatcherStrategy:
    type: None
````

**Comment by simonpasquier** [comment]
Nope it isn't possible: the main reason being that the AlertmanagerConfig object is self contained or we'd have to expose an AlertmanagerReceiver CRD that would be referenced within an AlertmanagerConfig object. But I feel that it complicates things more than necessary.
